File: 0552-student-attendance-record-ii/0552-student-attendance-record-ii.cpp
class Solution {
public:
    int checkRecord(int n) {
        vector<vector<int>> dp(2, vector<int>(3, 0));
        dp[0][0] = 1;
        int kMod = 1e9 + 7;
        int ans = 0;
        
        for(int i = 1; i <= n; ++i) {
            // if p
            vector<vector<int>> tmp(2, vector<int>(3, 0));
            for(int j = 0; j < 2; ++j)
                for(int k = 0; k < 3; ++k) 
                    tmp[j][0] = (tmp[j][0] + dp[j][k]) % kMod;
            
            // if a
            for(int k = 0; k < 3; ++k) tmp[1][0] = (tmp[1][0] + dp[0][k]) % kMod;
            for(int j = 0; j < 2; ++j)
                for(int k = 1; k < 3; ++k) tmp[j][k] = (tmp[j][k] + dp[j][k - 1]) % kMod;
            
            dp = tmp;
        }
        
        for(int j = 0; j < 2; ++j)
            for(int k = 0; k < 3; ++k)
                ans = (ans + dp[j][k]) % kMod;
    
        return ans;
        
    }
};

File: 0552-student-attendance-record-ii/0552-student-attendance-record-ii.java
class Solution {
    public int checkRecord(int n) {
        int ans = 0;
        int[][][] dp = new int[n + 1][2][3];
        int mod = 1_000_000_007;
        dp[0][0][0] = 1;
        for(int i = 1; i <= n; ++i) {
            // if present
            for(int j = 0; j < 2; ++j)
                for(int k = 0; k < 3; ++k) {
                    dp[i][j][0] = (dp[i][j][0] + dp[i - 1][j][k]) % mod;
                }
            // if absent
            for(int k = 0; k < 3; ++k) dp[i][1][0] = (dp[i][1][0] + dp[i - 1][0][k]) % mod;
            // if late
            for(int j = 0; j < 2; ++j)
                for(int k = 1; k < 3; ++k) {
                    dp[i][j][k] = (dp[i][j][k] + dp[i - 1][j][k - 1]) % mod;
                }
        }
        
        for(int j = 0; j < 2; ++j)
            for(int k = 0; k < 3; ++k)
                ans = (ans + dp[n][j][k]) % mod;
        
        return ans;
    }
}